### DISCUSSION TITLE
ignore crate: New API for custom ignore file processing?

### DISCUSSION THREAD

**Discussion by typesanitizer** [BurntSushi/ripgrep#2352] [General]
There is this API for adding custom ignore files.

```rust
    pub fn add_custom_ignore_filename<S: AsRef<OsStr>>(
        &mut self,
        file_name: S,
    ) -> &mut WalkBuilder {
```

However, it assumes that the file has the same syntax as a `.gitignore`. I'd like to be able to use `.gitattributes` like so:

```
glob mytool-skip=true
```

.gitattributes is recognized by `git check-attr`, making it more convenient than a custom ignore file. It is also used by GitHub (e.g. for marking generated files), so I'd like to use it instead of creating another filename `.mytoolignore` like `.rgignore`.

Would you be open to adding a new method which allows for a custom way to map the file contents to a list of git-like glob patterns?

The most ~hacky~ simple solution would probably be to have an extra function argument which does a (byte)string -> (byte)string transform on the file contents, with the return value being in `.gitignore` format. Doesn't feel terribly elegant, but it avoids the need to expose extra internal types. I saw a comment in the source code that the current APIs are too complicated to expose the core Ignore type directly. https://sourcegraph.com/github.com/BurntSushi/ripgrep/-/blob/crates/ignore/src/dir.rs?L11-14

**Comment by BurntSushi** [comment]
Could you please provide a simple concrete example of how you expect it to work? Show the inputs, show the code, show the desired output. It's really really hard to understand what you're asking for otherwise.

**Reply by typesanitizer** [reply]
Sorry, I should've given more concrete details up front. Here's is the structure I would like: (my tool would be working in this repository)

```
  .gitattributes
  myfile
  unrelatedfile
  .git/
```

Contents of `.gitattributes`

```
myfile mytool-skip=true
unrelatedfile linguist-generated=true
```

In code, the API would be invoked as something like:

```
fn map_gitattributes_to_gitignore(file_contents: Vec<u8>) -> Vec<u8> {
  // Identifies the lines which have mytool-skip=true and transforms:
  //
  // input:
  // myfile mytool-skip=true 
  // unrelatedfile linguist-generated=true
  // 
  // output: (valid .gitignore)
  // myfile
}

fn my_tool_main(root_path: PathBuf) {
  let walker = ignore::WalkBuilder::new(root_path)
    .add_custom_ignore_file_format(".gitattributes", map_gitattributes_to_gitignore)
    // like add_custom_ignore_filename, this will read the ".gitattributes" file
    // but will pass the data to the provided function, and interpret the return
    // value as a .gitignore file
    .build_parallel();
  // Do stuff with walker.
}
```

Does that help explain what I'm asking for? I'm not particularly tied to specific types/names for the API.
